etcdserver/etcdhttp/client_security.go

```diff
@@ -0,0 +1,399 @@
+	}
+}
+
+func (sh *securityHandler) forUser(w http.ResponseWriter, r *http.Request, user string) {
+	if !allowMethod(w, r.Method, "GET", "PUT", "DELETE") {
+		return
+	}
+	if !hasRootAccess(sh.sec, r) {
+		writeNoAuth(w)
+		return
+	}
+	w.Header().Set("X-Etcd-Cluster-ID", sh.clusterInfo.ID().String())
+
+	switch r.Method {
+	case "GET":
+		w.Header().Set("Content-Type", "application/json")
+		u, err := sh.sec.GetUser(user)
+		if err != nil {
+			writeError(w, err)
+			return
+		}
+		u.Password = ""
+		err = json.NewEncoder(w).Encode(u)
+		if err != nil {
+			log.Println("etcdhttp: forUser error encoding on", r.URL)
+			return
+		}
+	case "PUT":
+		var u security.User
+		err := json.NewDecoder(r.Body).Decode(&u)
+		if err != nil {
+			writeError(w, httptypes.NewHTTPError(http.StatusBadRequest, err.Error()))
+			return
+		}
+		if u.User != user {
+			httptypes.NewHTTPError(400, "User JSON name does not match the name in the URL")
+		}
+		err = sh.sec.CreateUser(u)
+		if err != nil {
+			writeError(w, err)
+			return
+		}
+	case "DELETE":
+		err := sh.sec.DeleteUser(user)
+		if err != nil {
+			writeError(w, err)
+			return
+		}
+	}
+}
+
+func (sh *securityHandler) actionUser(w http.ResponseWriter, r *http.Request, user string, action string) {
+	if !allowMethod(w, r.Method, "PUT") {
+		return
+	}
+	if !hasRootAccess(sh.sec, r) {
+		writeNoAuth(w)
+		return
+	}
+	w.Header().Set("X-Etcd-Cluster-ID", sh.clusterInfo.ID().String())
+	w.Header().Set("Content-Type", "application/json")
+	var in security.User
+	err := json.NewDecoder(r.Body).Decode(&in)
+	if err != nil {
+		writeError(w, httptypes.NewHTTPError(http.StatusBadRequest, err.Error()))
+		return
+	}
+	var out security.User
+	switch action {
+	case "grant":
+		out, err = sh.sec.UpdateUser(security.User{User: user, Grant: in.Grant})
+	case "revoke":
+		out, err = sh.sec.UpdateUser(security.User{User: user, Revoke: in.Revoke})
+	case "password":
+		out, err = sh.sec.UpdateUser(security.User{User: user, Password: in.Password})
+	default:
+		return
+	}
+	if err != nil {
+		writeError(w, err)
+		return
+	}
+	out.Password = ""
+	err = json.NewEncoder(w).Encode(out)
+	if err != nil {
+		log.Println("etcdhttp: actionUser error encoding on", r.URL)
+	}
+}
+
+func (sh *securityHandler) enableDisable(w http.ResponseWriter, r *http.Request) {
+	if !allowMethod(w, r.Method, "GET", "PUT", "DELETE") {
+		return
+	}
+	if !hasWriteRootAccess(sh.sec, r) {
+		writeNoAuth(w)
+		return
+	}
+	w.Header().Set("X-Etcd-Cluster-ID", sh.clusterInfo.ID().String())
+	w.Header().Set("Content-Type", "application/json")
+	isEnabled := sh.sec.SecurityEnabled()
+	switch r.Method {
+	case "GET":
+		err := json.NewEncoder(w).Encode(isEnabled)
+		if err != nil {
+			log.Println("etcdhttp: error encoding security state on", r.URL)
+		}
+	case "PUT":
+		var in security.User
+		err := json.NewDecoder(r.Body).Decode(&in)
+		if err != nil {
+			writeError(w, httptypes.NewHTTPError(http.StatusBadRequest, err.Error()))
```

Comment:
i guess we should not return the error directly. for example i can see this kind of message:
```
{"message":"json: cannot unmarshal bool into Go value of type security.User"}
```

```
{"message":"EOF"}
```